File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/infrastructure/SqliteGiftCardRepository.java
package it.giannibombelli.wsc2026.giftcard.infrastructure;

import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import org.jooq.DSLContext;
import org.jooq.Record2;
import org.jooq.Result;
import org.jooq.SQLDialect;
import org.jooq.impl.DSL;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.Optional;
import java.util.UUID;

import static org.jooq.impl.DSL.field;
import static org.jooq.impl.DSL.table;

public final class SqliteGiftCardRepository implements GiftCardRepository {
    private static final org.jooq.Table<?> GIFT_CARD = table("gift_card");
    private static final org.jooq.Field<String> ID = field("id", String.class);
    private static final org.jooq.Field<BigDecimal> BALANCE = field("balance", BigDecimal.class);

    private final DSLContext dsl;

    public SqliteGiftCardRepository(DataSource dataSource) {
        this.dsl = DSL.using(dataSource, SQLDialect.SQLITE);
    }

    @Override
    public void save(GiftCard giftCard) {
        String idStr = giftCard.id().value().toString();
        BigDecimal bal = giftCard.balance().value();

        // Upsert (PK exists → update; else insert) — deliberate simplicity for the workshop vertical.
        int updated = dsl.update(GIFT_CARD)
            .set(BALANCE, bal)
            .where(ID.eq(idStr))
            .execute();

        if (updated == 0) {
            dsl.insertInto(GIFT_CARD)
                .columns(ID, BALANCE)
                .values(idStr, bal)
                .execute();
        }
    }

    @Override
    public Optional<GiftCard> findById(GiftCardId id) {
        String idStr = id.value().toString();

        Result<Record2<String, BigDecimal>> rows = dsl
            .select(ID, BALANCE)
            .from(GIFT_CARD)
            .where(ID.eq(idStr))
            .fetch();

        if (rows.isEmpty()) {
            return Optional.empty();
        }

        Record2<String, BigDecimal> r = rows.getFirst();
        GiftCardId cardId = new GiftCardId(UUID.fromString(r.get(ID)));
        BigDecimal balance = r.get(BALANCE);

        return Optional.of(new GiftCard(cardId, new Money(balance)));
    }
}

File: typescript/src/giftcard/infrastructure/sqliteGiftCardRepository.ts
// Adapter repository per SQLite usando Drizzle ORM.

import Database from 'better-sqlite3';
import { drizzle } from 'drizzle-orm/better-sqlite3';
import type { BetterSQLite3Database } from 'drizzle-orm/better-sqlite3';
import { eq } from 'drizzle-orm';

import { Money } from '@/common/domain/primitive/money.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { GiftCard } from '../domain/giftcard/giftCard.js';
import { GiftCardId } from '../domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '../domain/ports/giftCardRepository.js';
import { giftCards } from './schema.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class SqliteGiftCardRepository implements GiftCardRepository {
  private readonly _db: BetterSQLite3Database;

  constructor(database: Database.Database) {
    requireDependency(database, "database");
    this._db = drizzle(database);
  }

  save(giftCard: GiftCard): void {
    const id = giftCard.id().value.value;
    const balance = giftCard.balance().value;

    this._db
      .insert(giftCards)
      .values({ id, balance })
      .onConflictDoUpdate({
        target: giftCards.id,
        set: { balance },
      })
      .run();
  }

  findById(id: GiftCardId): GiftCard | null {
    const rows = this._db
      .select()
      .from(giftCards)
      .where(eq(giftCards.id, id.value.value))
      .all();

    if (rows.length === 0) {
      return null;
    }

    const row = rows[0];
    return new GiftCard(new GiftCardId(Uuid.fromString(row.id)), new Money(row.balance));
  }
}
